### ISSUE TITLE
Numeric DecodeBinary Hangs

### ISSUE THREAD

**Issue by jimtsao** [jackc/pgtype#133]
The pg [Numeric Documentation](https://www.postgresql.org/docs/current/datatype-numeric.html) specifies a range "up to 131072 digits before the decimal point; up to 16383 digits after the decimal point".

I can encode a numeric value in this range successfully with the below:
```
max := new(big.Int).Exp(big.NewInt(10), big.NewInt(147454), nil)
max.Add(max, big.NewInt(1))
pgtype.Numeric{Int: max, Exp: -16383, Status: pgtype.Present}
```

However, when decoding and scanning it back from the database it seems to spend an inordinate amount of time in this loop (I haven't been able to wait it out to see how long it takes exactly).

https://github.com/jackc/pgtype/blob/5cb98120c10c0ed3125786007d6c0861d34b8f79/numeric.go#L527-L536

Running the query with a `context.WithTimeout` doesn't help stop the query and return early either. This could potentially be a security issue too, if someone is able to push through a large numeric input such as this and cause a server to hang when retrieving it.

In case anyone wants to optimize this part, you can peruse through the [source code](https://doxygen.postgresql.org/backend_2utils_2adt_2numeric_8c.html) for explanation on the numeric binary representation parts.

Any suggestions on how to optimise the binary decoding for large numerics?

**Comment by jackc** [comment]
The underlying problem was the ndigits field is an `uint16` and it was being read as a `int16`. An overflow would lead to negative number of digits. This eventually would lead to an infinite loop.

Fixed in a29019d.

**Comment by jimtsao** [comment]
Good catch. Thank you for the quick fix.
